presto-main/src/main/java/com/facebook/presto/cost/HistoricalPlanStatisticsUtil.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.spi.statistics.HistoricalPlanStatistics;
+import com.facebook.presto.spi.statistics.PlanStatistics;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import static java.lang.Double.isNaN;
+
+public class HistoricalPlanStatisticsUtil
+{
+    // Store historical data for last ``MAX_LAST_RUNS`` runs at maximum.
+    private static final int MAX_LAST_RUNS = 10;
+    // Maximum number of input tables till we store data of multiple last runs. Once number of input tables becomes
+    // larger than this, we store data of fewer historical runs. See getMaxLastRuns() function.
+    private static final int MAX_INPUT_TABLES = 100;
+    // Use a historical statistic when size of input tables in within 1 +/- ``THRESHOLD`` size range.
+    private static final double THRESHOLD = 0.1;
+
+    private HistoricalPlanStatisticsUtil() {}
+
+    /**
+     * Returns predicted plan statistics depending on historical runs
+     */
+    public static PlanStatistics getPredictedPlanStatistics(
+            HistoricalPlanStatistics historicalPlanStatistics,
+            List<PlanStatistics> inputTableStatistics)
+    {
+        Optional<PlanStatistics> overriddenStatistics = historicalPlanStatistics.getOverriddenStatistics();
+        Optional<List<PlanStatistics>> lastRunsStatistics = historicalPlanStatistics.getLastRunsStatistics();
+        Optional<List<List<PlanStatistics>>> lastRunsInputStatistics = historicalPlanStatistics.getLastRunsInputStatistics();
+
+        if (overriddenStatistics.isPresent()) {
+            return overriddenStatistics.get();
+        }
+
+        if (!lastRunsStatistics.isPresent() || !lastRunsInputStatistics.isPresent()) {
+            return PlanStatistics.empty();
+        }
+
+        if (lastRunsInputStatistics.get().size() != lastRunsStatistics.get().size() ||
```

Comment:
These seem like things that shouldn't happen.  Should we throw an exception in these cases instead?